src/lib/viewers/doc/BoxExcel/charts.js

```diff
@@ -0,0 +1,237 @@
+    /* eslint-disable */
+    /**
+     * sheet spreadsheet data
+     * themeColors array of colors
+     *
+     * @type {Object}
+     */
+    static propTypes = {
+        sheet: PropTypes.object.isRequired,
+        zoom: PropTypes.number,
+        themeColors: PropTypes.array.isRequired,
+        scrollLeftOffset: PropTypes.number.isRequired,
+        scrollTopOffset: PropTypes.number.isRequired
+    };
+    /* eslint-enable */
+
+    /**
+     * [constructor]
+     *
+     * @param {Object} props [description]
+     */
+    constructor(props) {
+        super(props);
+
+        const { sheet } = this.props;
+        this.state = {
+            charts: sheet['!charts'].map(this._parseChart)
+        };
+    }
+
+    /**
+     * Helper function of _parseChart
+     * Re-organize the data of a subplot
+     *
+     * @param  {Object} chart   Chart object parsed from sheetjs
+     * @param  {Array} subPlot  'cat' means chart label, 'val' means chart data
+     * @param  {number} counter counter of the whole chart
+     * @return {Object}         labels and data of a subplot, and updated counter
+     */
+    _getRawData = (chart, subPlot, counter) => {
+        /* global XLSX */
+        const { e } = XLSX.utils.decode_range(chart['!ref']);
+        const rowCount = e.r + 1;
+        let subLabels = [];
+        const subData = [];
+        let newCounter = counter;
+        subPlot.forEach((key) => {
+            for (let r = 0; r < rowCount; ++r) {
+                const pos = XLSX.utils.encode_cell({
+                    c: newCounter,
+                    r
+                });
+                if (key === 'cat') {
+                    subLabels.push(chart[pos].v);
+                } else if (key === 'val') {
+                    subData.push(chart[pos].v);
+                }
+            }
+            newCounter += 1;
+        });
+        if (!subLabels.length) subLabels = Array.from({ length: subData.length }, (_, i) => i + 1);
+
+        return { subLabels, subData, newCounter };
+    };
+
+    /**
+     * Whether the chart should display Axes
+     *
+     * @param  {string} type chart type string
+     * @return {boolean}     true for certain types
+     */
+    _displayAxes = (type) => {
+        return !(type === 'pie' || type === 'doughnut');
+    };
+
+    /**
+     * Re-organize sheetjs data to chartjs format
+     *
+     * @param  {Object} chart chart object parsed from sheetjs
+     * @return {Object}       chartjs needed properties
+     */
+    _parseChart = (chart) => {
+        const chartType = chart['!plot'][0].t;
+        const position = chart['!pos'];
+        let title = chart['!title'];
+        const legendPos = chart['!legend'] ? legendMap[chart['!legend'].pos] : null;
+        let counter = 0;
+
+        const rawDatasets = [];
+        const yAxes = [];
+        const plots = chart['!plot'];
+        let subPlots = [];
+        plots.forEach((plot, i) => {
+            subPlots = plot.ser;
```

Comment:
Refactor this into some util methods so it's clearer to understand what's going on